src/query/core/buildInitiate.ts

```diff
@@ -79,18 +80,98 @@ type StartMutationActionCreator<
 export type MutationActionCreatorResult<
   D extends MutationDefinition<any, any, any, any>
 > = Promise<
-  ReturnType<
-    BaseQueryResult<
-      D extends MutationDefinition<any, infer BaseQuery, any, any>
-        ? BaseQuery
-        : never
-    >
-  >
+  | { data: ResultTypeFrom<D> }
+  | {
+      error:
+        | BaseQueryError<
+            D extends MutationDefinition<any, infer BaseQuery, any, any>
+              ? BaseQuery
+              : never
+          >
+        | SerializedError
+    }
 > & {
```

Comment:
@phryneas I've tried fixing the type for the resolved value of the 'trigger' promise here, but the actual `error` type produced has an additional `undefined` type attached to it.

![image](https://user-images.githubusercontent.com/22929669/115955605-e266b480-a53a-11eb-975f-32aea184c26f.png)

Can you make any sense out of that?

I've temporarily suppressed the error in order to allow codesandbox to build, in order to see how it behaved there:
https://codesandbox.io/s/vanilla-typescript-forked-duy39?file=/src/index.ts

Reply:
Ah it's just occurred to me; would that be caused by the scenario where the `extends MutationDefinition` ternary fails?

```ts
        | BaseQueryError<
            D extends MutationDefinition<any, infer BaseQuery, any, any>
              ? BaseQuery
              : never
          >
```

Reply:
I'm particularly stumped because it looks like `BaseQueryError` is explicitly intended to *remove* the case with `{ error: undefined }`
```ts
export type BaseQueryError<BaseQuery extends BaseQueryFn> = Exclude<
  UnwrapPromise<ReturnType<BaseQuery>>,
  { error: undefined }
>['error']
```

Reply:
I pushed a new version which explicitly removes `undefined` from the type returned by `BaseQueryError` like so:

```ts
Promise<
  | { data: ResultTypeFrom<D> }
  | {
      error:
        | Exclude<
            BaseQueryError<
              D extends MutationDefinition<any, infer BaseQuery, any, any>
                ? BaseQuery
                : never
            >,
            undefined
          >
        | SerializedError
    }
>
```

But I'm iffy about two things:
- Is `undefined` actually supposed to be there for some reason?
- If not, and I remove it doing the above, and someone defines their error type like: `{ error: undefined }`, then the promise type will be wrong (Because it will remove `undefined`)